scylla-cql/src/types/serialize/row.rs

```diff
@@ -50,22 +55,310 @@ pub trait SerializeRow {
+            // Unfortunately, column names aren't guaranteed to be unique.
+            // We need to track not-yet-used columns in order to see
+            // whether some values were not used at the end, and report an error.
+            let mut unused_columns: HashSet<&str> = self.keys().map(|k| k.as_ref()).collect();
+
+            for col in ctx.columns.iter() {
+                match self.get(col.name.as_str()) {
+                    None => {
+                        return Err(mk_typck_err::<Self>(
+                            BuiltinTypeCheckErrorKind::MissingValueForColumn {
+                                name: col.name.clone(),
+                            },
+                        ))
+                    }
+                    Some(v) => {
+                        <T as SerializeCql>::serialize(v, &col.typ, writer.make_cell_writer())
+                            .map_err(|err| {
+                                mk_typck_err::<Self>(
+                                    BuiltinTypeCheckErrorKind::ColumnTypeCheckFailed {
+                                        name: col.name.clone(),
+                                        err,
+                                    },
+                                )
+                            })?;
+                        let _ = unused_columns.remove(col.name.as_str());
+                    }
+                }
+            }
+
+            if !unused_columns.is_empty() {
+                // Report the lexicographically first value for deterministic error messages
+                let name = unused_columns.iter().min().unwrap();
+                return Err(mk_typck_err::<Self>(
+                    BuiltinTypeCheckErrorKind::ColumnMissingForValue {
+                        name: name.to_string(),
+                    },
+                ));
+            }
+
+            Ok(())
+        }
+    };
+}
+
+impl<T: SerializeCql> SerializeRow for BTreeMap<String, T> {
+    impl_serialize_row_for_map!();
+}
+
+impl<T: SerializeCql> SerializeRow for BTreeMap<&str, T> {
+    impl_serialize_row_for_map!();
+}
+
+impl<T: SerializeCql, S: BuildHasher> SerializeRow for HashMap<String, T, S> {
+    impl_serialize_row_for_map!();
+}
+
+impl<T: SerializeCql, S: BuildHasher> SerializeRow for HashMap<&str, T, S> {
+    impl_serialize_row_for_map!();
+}
+
+impl<T: SerializeRow> SerializeRow for &T {
+    fn preliminary_type_check(ctx: &RowSerializationContext<'_>) -> Result<(), SerializationError> {
+        <T as SerializeRow>::preliminary_type_check(ctx)
     }
 
     fn serialize<W: RowWriter>(
         &self,
         ctx: &RowSerializationContext<'_>,
         writer: &mut W,
     ) -> Result<(), SerializationError> {
-        serialize_legacy_row(self, ctx, writer)
+        <T as SerializeRow>::serialize(self, ctx, writer)
     }
 }
 
+impl SerializeRow for SerializedValues {
+    fallback_impl_contents!();
+}
+
+impl<'b> SerializeRow for Cow<'b, SerializedValues> {
+    fallback_impl_contents!();
+}
```

Comment:
Is this a implementation created to make transition to new API easier? If so, let's add a TODO / create and issue so that we don't forget to remove it later.